packages/lesswrong/components/ultraFeed/UltraFeedObserver.tsx

```diff
@@ -0,0 +1,199 @@
+import React, {
+  createContext,
+  useContext,
+  useRef,
+  useEffect,
+  useCallback,
+  ReactNode,
+} from 'react';
+import { useCreate } from '../../lib/crud/withCreate';
+import { useCurrentUser } from "../common/withUser";
+
+type DocumentType = 'post' | 'comment' | 'spotlight';
+
+interface ObserveData {
+  documentId: string;
+  documentType: DocumentType;
+  postId?: string; // Relevant for comments
+}
+
+interface TrackExpansionData {
+  documentId: string;
+  documentType: 'post' | 'comment';
+  postId?: string;
+  level: number;
+  maxLevelReached: boolean;
+  wordCount: number;
+}
+
+interface UltraFeedObserverContextType {
+  observe: (element: Element, data: ObserveData) => void;
+  unobserve: (element: Element) => void;
+  trackExpansion: (data: TrackExpansionData) => void;
+}
+
+const UltraFeedObserverContext = createContext<UltraFeedObserverContextType | null>(null);
+
+const VIEW_THRESHOLD_MS = 500;
+const INTERSECTION_THRESHOLD = 0.5; // 50% visible
+
+const mapDocumentTypeToCollectionName = (documentType: DocumentType): "Posts" | "Comments" | "Spotlights" => {
```

Comment:
am confused about why this needs to be a function instead of just:
```
const documentTypeToCollectionName = {
  post: "Posts",
  comment: "Comments",
  spotlight: "Spotlights",
};
```

If you want to be sure you don't mess up the types you can stick a `satisfies Record<DocumentType, "Posts" | "Comments" | "Spotlights">;` after it, which avoids erasing the actually mapping.